Feat/togglegroup

azeez1776 opened the pull request:
:star: Feat: ToggleGroup Feature Closes #607 

- Added the ToggleGroup Component.
- Added Documentation for the component.
- No using Radix libraries.

natew commented:
We need to be in sync as well with everyone - right now we have a variety of PRs and I think this overlaps with RadioGroup in some ways, I haven't had time to review everything but please do review and make sure things aren't overlapping.

azeez1776 commented:
Sure, will review and make sure to fix where there is an overlapping.

azeez1776 commented:
- Added Roving Focus feature to the toggle group(Full keyboard navigation).
- Supports single and multiple pressed buttons.
- Supports horizontal or vertical orientation.

### Video below shows how toggle groups responds to both the keyboard navigations and mouse clicks.
https://user-images.githubusercontent.com/39125180/218272673-8d6bed0d-262e-4319-a3c8-26953adeb2ef.mp4

azeez1776 commented:
Hello @natew , I think toggle group works fine now. Any suggestions for improvement? or is it okay right now.

natew commented:
The middle items not getting the right borders is unfortunate, we'll have to figure that out or combine with XGroup and fix up groups too...

azeez1776 commented:
That's true the middle item is off, I tried accessing the first and last child but was unfortunate, I will try using the XGroup to solve this issue.

azeez1776 commented:
Resolved the border radius on the middle toggle item. The border radius appears on the end corners of the toggle group.
                        <img width="513" alt="Screenshot_20230213_082207" src="https://user-images.githubusercontent.com/39125180/218528294-968e8912-01ff-4cc2-a130-bb63e3040a16.png">

natew commented:
outline showing on safari:

<img width="308" alt="CleanShot 2023-02-13 at 14 08 44@2x" src="https://user-images.githubusercontent.com/12100/218603399-ee4162f0-fa3b-49d7-906d-92361c9c1e50.png">

Also thinking about this one, handling the border radius seems outside ToggleGroup, it should mostly be without any styles by default and if you want just use XGroup with it. If we can demo it with XGroup that'd be ideal so the demo looks nice and shows how to do it right.

natew commented:
Active state should use theme="active", use elevation for a nice shadow on the ToggleGroup etc.

natew commented:
We have some stuff coming in the PR https://github.com/tamagui/tamagui/pull/705 that should dovetail here.

natew commented:
@azeez1776 sorry this is taking a bit, we merged the new Group stuff so you should be good to use that for some of your work here!

azeez1776 commented:
No worries @natew , I will start working on it as soon as possible.

azeez1776 commented:
Made the following changes:
- Added elevation and border radius( border radius not part of togglegroup component  as before)  on the demo togglegroup.
- Tried fixing the outline issue on safari.
- Used theme="active" on selected toggle.
- Added two togglegroup sizes of $3 and $4 on the demo.
- Used XGroup on the demo which helped on the border radius of the first and last child.

azeez1776 commented:
https://user-images.githubusercontent.com/39125180/223794732-a5edd105-1222-472b-8676-b5beceb5a0ca.mp4

azeez1776 commented:
I think something broke, the border radius from the XGroup I used on the demo was working fine until I pulled the recent changes.
### It looked like this before I pulled the recent changes.
<img width="548" alt="Screenshot_20230308_072838" src="https://user-images.githubusercontent.com/39125180/223803286-0c33db4d-43a6-4d47-90d7-479a616efab9.png">


### After I pulled and merged the recent changes it looks like this

<img width="668" alt="Screenshot_20230308_093532" src="https://user-images.githubusercontent.com/39125180/223804450-9d9471b2-995d-48b0-ac06-81c3e9399eab.png">

alitnk commented on packages/demos/src/ToggleGroupDemo.tsx:
```diff
@@ -0,0 +1,50 @@
+          <ToggleGroup.Item size={props.size} value="right" aria-label="Right aligned">
+            <AlignRight />
+          </ToggleGroup.Item>
+        </XGroup>
```
We *could* have `Group` baked into `ToggleGroup`, like how Tabs works, where `Tabs.List` is a Group and `Tabs.Trigger` implements `Group.Item` using the `useGroupItem` hook.

I'm suggesting this since ToggleGroup is technically a group (even has it in its name) and users will probably always use it with a Group. So from something like:
```tsx
<ToggleGroup>
 <XGroup>
   <XGroup.Item> // optional - would fallback to old impl of group items
     <ToggleGroup.Item>
       <AlignLeft />
     </ToggleGroup.Item>
   </XGroup.Item>
 </XGroup>
</ToggleGroup/>
```
to: 
```tsx
<ToggleGroup> // extends Group
  <ToggleGroup.Item> // Group.Item functionality through useGroupItem()
    <AlignLeft />
  </ToggleGroup.Item>
</ToggleGroup/>
```

I guess the only difference on ToggleGroup's default behavior would be the appliance of border-radius, which users can disable by passing `disablePassBorderRadius`. I might be missing something, though.

What do you think, @natew?

alitnk replied:
Great! I'd still hold it off to see what Nate thinks as he might not agree with this proposal.

azeez1776 replied:
Alright, let wait

natew commented:
Sorry it's taken me a while to get back on this PR. It looks like great work though, if we can make that last change and fix conflicts I'll give it a thorough look at and fix anything I see, and aim to get it out this weekend.

azeez1776 commented:
No worries Nate, I will make the change today, hopefully I will finish before the end of the day.

azeez1776 commented:
![Screenshot_20230326_060227](https://user-images.githubusercontent.com/39125180/227789000-a8760879-906d-46be-9026-d67b3c325ab0.png)

natew commented:
We're finishing the work in another PR, I'll lyk when released
